### PR TITLE
Diff processing and evaluations

### PR DESCRIPTION
**Pull request by honglu2875**
The goal is to add some utility functions for diff processing and evaluations.
- [x] Add diff patching tools.
- [x] Add diff evaluations (see `verify_diff`).
- [x] Add `benchmark_diff.py` which evaluates the diff model accuracy of bug-fixing benchmark task of 4-parity. 
- [x] Add pytest to the CI since I really wrote a bunch of tests on diff patching and diff verifying (I won't say it's enough... Diff patching is a subtle thing and there might be more edge cases.) 

A note on diff patching:
- I'm trying to patch it in a relatively loose way (try not to reject the diff if there are correctible mistakes) because our goal was to generate diversity. A stricter format check is in `verify_diff` and it returns my own version of status codes. Please see the comments in the code for details.

Another minor change:
 - Exposed `Genotype` and `BaseEnvironment` for other people to inherit (could be convenient for [Architext's integration](https://github.com/CarperAI/ArchitextRL/pull/1), for example).

### CODE REVIEW

File: elm/Pipfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[[source]]
```
Comment (herbiebradley) [inline elm/Pipfile:1]: Do we actually need this file or is it an artifact of some env manager?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -24,10 +24,13 @@ jobs:
...
+          python -m pip install .[dev]
       - name: Lint with flake8
         run: |
           # Stop the build if there are Python syntax errors or undefined names
           flake8 . --count --select=E9,F63,F7 --show-source --statistics --max-line-length=88
           # Full lint, will not stop the build on failure.
           flake8 . --count --exit-zero --ignore=E203,E501,F811 --show-source --statistics --max-line-length=88
+      - name: Test with pytest
+        run: |
+          pytest tests/
```
Comment (herbiebradley) [inline .github/workflows/build.yml:36]: Think you can remove the `tests/` though it's a minor thing. Also make sure we have a newline at the end.

File: elm/benchmark_diff.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    Mutate code to create n bugs. Output the prompt in diff format.
+    Args:
+        n_bugs: number of bugs to introduce (from 1 to 5).
+        task: (Optional) the task to be performed.
+    Returns:
+        mutated_code, function_string
+    """
+    mutation_template = [
+        f"<NME> {task}.py\n<BEF> ",
+        "\n<MSG> Fixed bugs",
```
Comment (herbiebradley) [inline elm/benchmark_diff.py:65]: Are we not actually appending <DFF> to the prompt here?

File: requirements.txt [hunk 4]
```diff
@@ -1,10 +1,11 @@
 hydra-core>=1.2.0
 wandb>=0.13
-tree-sitter
-pygraphviz
-graphviz
 numpy
 transformers>=4.22.0
 tokenizers
+torch
```
Comment (herbiebradley) [inline requirements.txt:6]: AFAIK torch shouldn't be needed here since it's a dependency of transformers.

File: requirements.txt [hunk 5]
```diff
@@ -1,10 +1,11 @@
...
+swig
```
Comment (herbiebradley) [inline requirements.txt:7]: What is this for?

File: elm/util/diff_eval.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    """
+    Verify the validity of a complete diff text.
+
+    Args:
+        diff_text: the complete diff text.
+            The overall format conforms "<NME> ...\n<BEF> ...\n<MSG> ...\n<DFF> ..." and the text
+            after <DFF> has 1 or more lines of "@@ -x,y +a,b @@" followed by the corresponding hunk.
+    Returns:
+        a custom status code (binary):
+            0b000: valid diff.
```
Comment (herbiebradley) [inline elm/util/diff_eval.py:185]: Can we make the status code an Enum? Just seems a bit neater maybe?

**Review by herbiebradley — CHANGES_REQUESTED** [review]
Very nice, thanks! Just had a few questions!

File: elm/Pipfile (hunk 1 above)
Comment (honglu2875) [reply elm/Pipfile:1]: Looks like it came from Francisco's patch.
@TheExGenesis Looks like some kind of docker stuff. Any ideas?

File: elm/benchmark_diff.py (hunk 3 above)
Comment (honglu2875) [reply elm/benchmark_diff.py:65]: This was modified based on `benchmark.py` and I got confused about this line too (where was the code??) :joy:. Is this what you are asking? Turns out there is a `mutation_template.insert(1, func_str)` in line 76 that inserts the code to the middle. But maybe we can make it less confusing.

File: elm/benchmark_diff.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    print(OmegaConf.to_yaml(cfg))
+    print("-----------------  End -----------------")
+
+    device = torch.device("cuda" if cfg.cuda else "cpu")
+    config = AutoConfig.from_pretrained(cfg.model)
+    # Sometimes our model just fresh came out of training. Force use_cache to be true.
+    config.use_cache = True
+    tokenizer = AutoTokenizer.from_pretrained(cfg.model)
+    if cfg.gpus > 1:
+        model = torch.nn.DataParallel(
```
Comment (honglu2875) [inline elm/benchmark_diff.py:98]: Oh I want to make a note here that DataParallel doesn't work (even for 1 node). This applies to codegen util https://github.com/CarperAI/OpenELM/blob/281429b8b9f9ca955239c1ec617eb62bced280f5/elm/codegen/codegen_utilities.py#L111
as well. Better way (also recommended by pytorch) is to use DDP but that would make the code a little complicated.

File: requirements.txt (hunk 4 above)
Comment (honglu2875) [reply requirements.txt:6]: I remember I ran into the problem of missing torch when doing fresh install.. Let me double check.

File: requirements.txt (hunk 5 above)
Comment (honglu2875) [reply requirements.txt:7]: Without this `swig` (I don't know what it is), there will be an error from box2d
`error: command 'swig' failed: No such file or directory`.
Same goes for `box2d-py` which seems to be needed to avoid an error in sodaracer env about an environment variable or something.

There are a lot of strange stuff around this box2d thing.

File: elm/util/diff_eval.py (hunk 6 above)
Comment (honglu2875) [reply elm/util/diff_eval.py:185]: Yeah that's a good point!

File: elm/benchmark_diff.py (hunk 3 above)
Comment (herbiebradley) [reply elm/benchmark_diff.py:65]: Yeah I think I wrote the insert and now I've been confused by my own inherited code 😆

File: requirements.txt (hunk 4 above)
Comment (honglu2875) [reply requirements.txt]: @herbiebradley Oh I remember the problem now! In CI it reports the error
`ModuleNotFoundError: No module named 'torch'`
in pytest...

**Review by herbiebradley — APPROVED** [review]
LGTM, thanks for sorting out the Github Actions!
